src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -783,7 +785,21 @@ std::future<Coordination::GetResponse> ZooKeeper::asyncGet(const std::string & p
             promise->set_value(response);
     };
 
-    impl->get(path, std::move(callback), {});
+    impl->get(path, std::move(callback), watch_callback);
+    return future;
+}
+
+std::future<Coordination::GetResponse> ZooKeeper::tryAsyncGetNoThrow(const std::string & path, Coordination::WatchCallback watch_callback)
+{
+    auto promise = std::make_shared<std::promise<Coordination::GetResponse>>();
```

Comment:
I don't like shared_ptr here. It's confusing, because you could use promise object only once and set value to promise from only one thread. I really don't know what will be otherwise.  And a future-promise pair has already a shared state and with shared_ptr there will be two random accesses to the memory.. It won't be a bottleneck, but note about it.

You can just create promise on a stack and move it to the lambda.

Reply:
Yes, I've expected the same behavior, but ... https://stackoverflow.com/a/33437008